The build_and_test job failed with exactly the same error messages as I have seen recently in https://github.com/ros2/rosbag2/pull/1342
```
FAIL: test_record.TestRecordAfterShutdown.test_exit_code[sqlite3]
2023-06-03T19:45:38.7299964Z ----------------------------------------------------------------------------------------------------------------------------------
2023-06-03T19:45:38.7300321Z Traceback (most recent call last):
2023-06-03T19:45:38.7300670Z   File "/__w/rosbag2/rosbag2/ros_ws/src/s1oktrgw0j/rosbag2/ros2bag/test/test_record.py", line 73, in test_exit_code
2023-06-03T19:45:38.7301027Z     launch_testing.asserts.assertExitCodes(
2023-06-03T19:45:38.7301467Z   File "/__w/rosbag2/rosbag2/ros_ws/build/launch_testing/launch_testing/asserts/assert_exit_codes.py", line 62, in assertExitCodes
2023-06-03T19:45:38.7301967Z     assert info.returncode in allowable_exit_codes, 'Proc {} exited with code {}'.format(
2023-06-03T19:45:38.7302418Z AssertionError: Proc ros2bag-cli-7 exited with code -9
2023-06-03T19:45:38.7302839Z ----------------------------- Captured stdout call -----------------------------
2023-06-03T19:45:38.7303047Z 
2023-06-03T19:45:38.7303143Z Starting test run test_record[sqlite3]
2023-06-03T19:45:38.7303591Z [INFO] [launch]: All log files can be found below /github/home/.ros/log/2023-06-03-19-45-17-529104-0dd7204ee0ae-126136
2023-06-03T19:45:38.7303940Z [INFO] [launch]: Default logging verbosity is set to INFO
2023-06-03T19:45:38.7304291Z [INFO] [ros2bag-cli-7]: process started with pid [126220]
2023-06-03T19:45:38.7304840Z [ros2bag-cli-7] stdin is not a terminal device. Keyboard handling disabled.[INFO] [1685821526.257381155] [rosbag2_recorder]: Press SPACE for pausing/resuming
2023-06-03T19:45:38.7305542Z [ros2bag-cli-7] [INFO] [1685821526.260584248] [rosbag2_storage]: Opened database '/tmp/tmpcx5wk9g_/ros2bag_test_record/ros2bag_test_record_0.db3' for READ_WRITE.
2023-06-03T19:45:38.7306069Z [ros2bag-cli-7] [INFO] [1685821526.264542138] [rosbag2_recorder]: Event publisher thread: Starting
2023-06-03T19:45:38.7306561Z [ros2bag-cli-7] [INFO] [1685821526.265532736] [rosbag2_recorder]: Listening for topics...
2023-06-03T19:45:38.7307011Z [ros2bag-cli-7] [INFO] [1685821526.270718724] [rosbag2_recorder]: Subscribed to topic '/rosout'
2023-06-03T19:45:38.7307436Z [INFO] [ros2bag-cli-7]: sending signal 'SIGINT' to process[ros2bag-cli-7]
2023-06-03T19:45:38.7307965Z [ERROR] [ros2bag-cli-7]: process[ros2bag-cli-7] failed to terminate '5' seconds after receiving 'SIGINT', escalating to 'SIGTERM'
2023-06-03T19:45:38.7308438Z [INFO] [ros2bag-cli-7]: sending signal 'SIGTERM' to process[ros2bag-cli-7]
2023-06-03T19:45:38.7308974Z [ERROR] [ros2bag-cli-7]: process[ros2bag-cli-7] failed to terminate '10.0' seconds after receiving 'SIGTERM', escalating to 'SIGKILL'
2023-06-03T19:45:38.7309436Z [INFO] [ros2bag-cli-7]: sending signal 'SIGKILL' to process[ros2bag-cli-7]
2023-06-03T19:45:38.7310002Z [ERROR] [ros2bag-cli-7]: process has died [pid 126220, exit code -9, cmd 'ros2 bag record -a -s sqlite3 --output /tmp/tmpcx5wk9g_/ros2bag_test_record'].
2023-06-03T19:45:38.7310263Z 
2023-06-03T19:45:38.7310369Z Starting test run test_record[mcap]
2023-06-03T19:45:38.7310818Z [INFO] [launch]: All log files can be found below /github/home/.ros/log/2023-06-03-19-45-17-529104-0dd7204ee0ae-126136
2023-06-03T19:45:38.7311165Z [INFO] [launch]: Default logging verbosity is set to INFO
2023-06-03T19:45:38.7311511Z [INFO] [ros2bag-cli-8]: process started with pid [126237]
2023-06-03T19:45:38.7312041Z [ros2bag-cli-8] stdin is not a terminal device. Keyboard handling disabled.[INFO] [1685821536.633422675] [rosbag2_recorder]: Press SPACE for pausing/resuming
2023-06-03T19:45:38.7312563Z [ros2bag-cli-8] [INFO] [1685821536.638780051] [rosbag2_recorder]: Event publisher thread: Starting
2023-06-03T19:45:38.7312999Z [ros2bag-cli-8] [INFO] [1685821536.640277344] [rosbag2_recorder]: Listening for topics...
2023-06-03T19:45:38.7313438Z [ros2bag-cli-8] [INFO] [1685821536.645916818] [rosbag2_recorder]: Subscribed to topic '/rosout'
2023-06-03T19:45:38.7313859Z [INFO] [ros2bag-cli-8]: sending signal 'SIGINT' to process[ros2bag-cli-8]
2023-06-03T19:45:38.7314332Z [ros2bag-cli-8] [INFO] [1685821536.649572001] [rosbag2_recorder]: Subscribed to topic '/events/write_split'
2023-06-03T19:45:38.7314847Z [ros2bag-cli-8] [INFO] [1685821536.650399197] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
2023-06-03T19:45:38.7315323Z [ros2bag-cli-8] [INFO] [1685821536.651652191] [rosbag2_recorder]: Event publisher thread: Exiting
2023-06-03T19:45:38.7315982Z [INFO] [ros2bag-cli-8]: process has finished cleanly [pid 126237]
```
It looks like something got changed recently in underlying sources which is leading to those error.

--- a/rosbag2_py/src/rosbag2_py/_transport.cpp
+++ b/rosbag2_py/src/rosbag2_py/_transport.cpp
@@ -190,16 +190,23 @@ public:
 class Recorder
 {
 private:
-  std::unique_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
+  static std::unique_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
+  std::shared_ptr<rosbag2_transport::Recorder> recorder_;
 
 public:
   Recorder()
   {
-    rclcpp::init(0, nullptr);
+    auto init_options = rclcpp::InitOptions();
+    init_options.shutdown_on_signal = false;
+    rclcpp::init(0, nullptr, init_options, rclcpp::SignalHandlerOptions::None);
     exec_ = std::make_unique<rclcpp::executors::SingleThreadedExecutor>();
     std::signal(
       SIGTERM, [](int /* signal */) {
-        rclcpp::shutdown();
+        rosbag2_py::Recorder::cancel();
+      });
+    std::signal(
+      SIGINT, [](int /* signal */) {
+        rosbag2_py::Recorder::cancel();
       });
   }
 
@@ -218,23 +225,38 @@ public:
     }
 
     auto writer = rosbag2_transport::ReaderWriterFactory::make_writer(record_options);
-    auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+    recorder_ = std::make_shared<rosbag2_transport::Recorder>(
       std::move(writer), storage_options, record_options, node_name);
-    recorder->record();
+    recorder_->record();
 
-    exec_->add_node(recorder);
+    exec_->add_node(recorder_);
     // Release the GIL for long-running record, so that calling Python code can use other threads
     {
       py::gil_scoped_release release;
-      exec_->spin();
+      while (!exit_) {
+        exec_->spin_all(std::chrono::milliseconds(30));
+      }
+      if (recorder_) {
+        recorder_->stop();
+        // Need to spin once to be able to send notifications about bag split and close
+        exec_->spin_some(std::chrono::milliseconds(30));
+      }
     }
   }
 
-  void cancel()
+  static void cancel()
   {
-    exec_->cancel();
+    exit_ = true;
+    if (exec_) {
+      exec_->cancel();
+    }
   }
+
+  static std::atomic_bool exit_;
 };
+
+std::unique_ptr<rclcpp::executors::SingleThreadedExecutor> Recorder::exec_;
+std::atomic_bool Recorder::exit_{false};
 
 // Return a RecordOptions struct with defaults set for rewriting bags.
 rosbag2_transport::RecordOptions bag_rewrite_default_record_options()
@@ -360,7 +382,7 @@ PYBIND11_MODULE(_transport, m) {
   .def(
     "record", &rosbag2_py::Recorder::record, py::arg("storage_options"), py::arg("record_options"),
     py::arg("node_name") = "rosbag2_recorder")
-  .def("cancel", &rosbag2_py::Recorder::cancel)
+  .def_static("cancel", &rosbag2_py::Recorder::cancel)
   ;
 
   m.def(

--- a/rosbag2_py/src/rosbag2_py/_transport.cpp
+++ b/rosbag2_py/src/rosbag2_py/_transport.cpp
@@ -191,17 +191,7 @@ class Recorder
 {
 private:
   using ExecutorT = rclcpp::executors::SingleThreadedExecutor;
-  static std::unique_ptr<ExecutorT, void (*)(ExecutorT *)> & get_executor_instance()
-  {
-    static std::unique_ptr<ExecutorT, void (*)(ExecutorT *)> executor(
-      new ExecutorT(),
-      [](ExecutorT * exec_ptr) {
-        delete exec_ptr;
-        rclcpp::shutdown();
-      }
-    );
-    return executor;
-  }
+  std::unique_ptr<ExecutorT> exec_;
   std::shared_ptr<rosbag2_transport::Recorder> recorder_;
 
 public:
@@ -209,9 +199,9 @@ public:
   {
     auto init_options = rclcpp::InitOptions();
     init_options.shutdown_on_signal = false;
-    if (!rclcpp::ok()) {
-      rclcpp::init(0, nullptr, init_options, rclcpp::SignalHandlerOptions::None);
-    }
+    rclcpp::init(0, nullptr, init_options, rclcpp::SignalHandlerOptions::None);
+
+    exec_ = std::make_unique<ExecutorT>();
     std::signal(
       SIGTERM, [](int /* signal */) {
         rosbag2_py::Recorder::cancel();
@@ -222,7 +212,10 @@ public:
       });
   }
 
-  virtual ~Recorder() = default;
+  virtual ~Recorder()
+  {
+    rclcpp::shutdown();
+  }
 
   void record(
     const rosbag2_storage::StorageOptions & storage_options,
@@ -238,30 +231,27 @@ public:
       std::move(writer), storage_options, record_options, node_name);
     recorder_->record();
 
-    get_executor_instance()->add_node(recorder_);
+    exec_->add_node(recorder_);
     try {
       // Release the GIL for long-running record, so that calling Python code can use other threads
       py::gil_scoped_release release;
       while (!exit_) {
-        get_executor_instance()->spin_all(std::chrono::milliseconds(30));
+        exec_->spin_all(std::chrono::milliseconds(10));
       }
       if (recorder_) {
         recorder_->stop();
-        // Need to spin once to be able to send notifications about bag split and close
-        get_executor_instance()->spin_some(std::chrono::milliseconds(30));
+        // Need to spin node to be able to send notifications about bag split and close
+        exec_->spin_all(std::chrono::milliseconds(10));
       }
     } catch (...) {
-      get_executor_instance()->remove_node(recorder_);
+      exec_->remove_node(recorder_);
     }
-    get_executor_instance()->remove_node(recorder_);
+    exec_->remove_node(recorder_);
   }
 
   static void cancel()
   {
     exit_ = true;
-    if (get_executor_instance()) {
-      get_executor_instance()->cancel();
-    }
   }
 
   static std::atomic_bool exit_;

--- a/rosbag2_py/test/test_transport.py
+++ b/rosbag2_py/test/test_transport.py
@@ -80,6 +80,7 @@ def test_record_cancel(tmp_path, storage_id):
     metadata_io = rosbag2_py.MetadataIo()
     assert wait_for(lambda: metadata_io.metadata_file_exists(bag_path),
                     timeout=rclpy.duration.Duration(seconds=3))
+    record_thread.join()
 
     metadata = metadata_io.read_metadata(bag_path)
     assert(len(metadata.relative_file_paths))

